hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java

```diff
@@ -362,25 +353,26 @@ void tryStartJob(Function<Long, Long> executionIdSupplier) {
     }
 
     private void rewriteDagWithSnapshotRestore(DAG dag, long snapshotId) {
-        logger.info(jobIdString() + ": restoring state from snapshotId=" + snapshotId);
-        for (Vertex vertex : dag) {
-            // We add the vertex even in case when the map is empty: this ensures, that
-            // Processor.finishSnapshotRestore() method is always called on all vertices in
-            // a job which is restored from a snapshot.
-            String mapName = snapshotDataMapName(jobId, snapshotId, vertex.getName());
-            Vertex readSnapshotVertex = dag.newVertex(
-                    SNAPSHOT_VERTEX_PREFIX + "read." + vertex.getName(), readMapP(mapName)
-            );
-            Vertex explodeVertex = dag.newVertex(
-                    SNAPSHOT_VERTEX_PREFIX + "explode." + vertex.getName(), ExplodeSnapshotP::new
-            );
-
-            readSnapshotVertex.localParallelism(vertex.getLocalParallelism());
-            explodeVertex.localParallelism(vertex.getLocalParallelism());
-
-            int destOrdinal = dag.getInboundEdges(vertex.getName()).size();
-            dag.edge(between(readSnapshotVertex, explodeVertex).isolated())
-               .edge(new SnapshotRestoreEdge(explodeVertex, vertex, destOrdinal));
+        String mapName = jobRecord.successfulSnapshotDataMapName();
+        logger.info("State of " + jobIdString() + " will be restored from snapshot " + snapshotId + ", map=" + mapName);
+
+        List<Vertex> originalVertices = new ArrayList<>();
+        dag.iterator().forEachRemaining(originalVertices::add);
+
+        Map<String, Integer> vertexToOrdinal = new HashMap<>();
+        Vertex readSnapshotVertex = dag.newVertex(SNAPSHOT_VERTEX_PREFIX + "read",
```

Comment:
what will be the parallelism here? since we only have one vertex, the throughput risks being much lower?

Reply:
It will be the default, which is 2. 

The map reading is mostly io-bound. It reads the snapshot data chunks, which are `byte[]`, so they are easy to deserialize (`System.arraycopy`). The actual objects are deserialized in expand-snapshot-p, which is cooperative and uses max parallelism. Map reading uses `MapProxyImpl.iterator`, which is a sync operation, so there will be some waiting. Two threads will keep cpu and network busy enough.

However, this is just an "informed opinion", not based on benchmarks.